aksel.nav.no/website/pages/templates/skjemavalidering/demo.tsx

```diff
@@ -0,0 +1,175 @@
+  const [formSubmitted, setFormSubmitted] = React.useState(false);
+
+  const [personnummer, setPersonnummer] = React.useState("");
+  const [personnummerError, setPersonnummerError] = React.useState("");
+  const [transportmiddel, setTransportmiddel] = React.useState("");
+  const [transportmiddelError, setTransportmiddelError] = React.useState(false);
+
+  function onSubmit(event: React.FormEvent) {
+    event.preventDefault();
+    setHasTriedToSubmit(hasTriedToSubmit + 1);
+    let isValidForm = true;
+
+    const personnummerErrorMsg = validatePersonnummer(personnummer);
+    setPersonnummerError(personnummerErrorMsg);
+    if (personnummerErrorMsg) {
+      isValidForm = false;
+    }
+    const transportmiddelInvalid = !transportmiddel;
+    setTransportmiddelError(transportmiddelInvalid);
+    if (transportmiddelInvalid) {
+      isValidForm = false;
+    }
+
+    if (isValidForm) {
+      setFormSubmitted(true);
+    }
+  }
+
+  useEffect(() => {
+    errorSummaryRef.current?.focus();
+  }, [hasTriedToSubmit]);
+
+  const showErrorSummary = Boolean(
+    hasTriedToSubmit && (personnummerError || transportmiddelError),
+  );
+
+  if (formSubmitted)
+    return (
+      <Page style={{ marginTop: 20 }}>
+        <Page.Block as="main" width="lg" gutters>
+          <VStack gap="8" align="center">
+            <Heading size="large">Demo slutt</Heading>
+            <Button
+              onClick={() => {
+                location.hash = "";
+                location.reload();
+              }}
+            >
+              Nullstill
+            </Button>
+          </VStack>
+        </Page.Block>
+      </Page>
+    );
+
+  return (
+    <Page style={{ marginTop: 20 }}>
+      <Page.Block as="main" width="lg" gutters>
+        <form onSubmit={onSubmit}>
+          <VStack gap="8">
+            <TextField
+              id="personnummer"
+              label="Personnummer"
+              value={personnummer}
+              onChange={(e) => setPersonnummer(e.currentTarget.value)}
+              onBlur={() => {
+                Boolean(hasTriedToSubmit) &&
+                  setPersonnummerError(validatePersonnummer(personnummer));
+              }}
+              error={personnummerError}
+            />
+            <RadioGroup
+              id="transportmiddel"
+              tabIndex={-1}
+              legend="Transportmiddel"
+              value={transportmiddel}
+              onChange={(newValue) => {
+                setTransportmiddel(newValue);
+                setTransportmiddelError(false);
+              }}
+              error={transportmiddelError && "Du må velge et transportmiddel."}
+            >
+              <Radio value="car">Bil</Radio>
+              <Radio value="walking">Gange</Radio>
+              <Radio value="public">Kollektivtransport</Radio>
+            </RadioGroup>
+
+            {showErrorSummary && (
+              <ErrorSummary
+                ref={errorSummaryRef}
```

Comment:
Might be cooking up something to experimental, but would this be a better pattern than using useEffect?
```jsx
  useEffect(() => {
    errorSummaryRef.current?.focus();
  }, [hasTriedToSubmit]);
```

```
  const shouldFocusErrorSummary = React.useRef(true);

   ...
 <ErrorSummary
    ref={(ref) => {
      shouldFocusErrorSummary.current && ref?.focus();
      shouldFocusErrorSummary.current = false;
    }}
    heading="Du må fikse disse feilene før du kan fortsette:"
  >
```

Seems to be working great while testing, but might have forgotten some edgecases.

Wil autofocus when mounted for the first time only here. Could still add the ref to `const errorSummaryRef = React.useRef<HTMLDivElement>(null);` if on wants to manually focus it again while mounted

Reply:
While the code might be a little more complex, i still think we can strive to follow react best practices of avoiding `useEffect` where possible. https://react.dev/learn/you-might-not-need-an-effect

I tried another concept avoiding any extra refs:
```tsx

function focusErrorSummary() {
  errorSummaryRef.current?.focus();
}

 if (Object.values(newErrors).some(Boolean)) {
  setFormState({ ...formState, tries: formState.tries + 1 });
  focusErrorSummary();
} else {
  setFormState({ ...formState, submitted: true });
}

<ErrorSummary
  ref={(ref) => {
    if (!errorSummaryRef.current) {
      errorSummaryRef.current = ref;
      focusErrorSummary();
      return;
    }
    errorSummaryRef.current = ref;
  }}
  heading="Du må rette disse feilene før du kan fortsette:"
>
```

`focusErrorSummary` or a variation of it allows us to take a more imperative approach where the user tells the app when to focus, more alike the `react-hook-forms` demo where focus is called `onSubmit`

Reply:
The page you linked to does say that "Code that runs because a component was displayed should be in Effects", which I thought would apply to this use case, but the imperative approach might be clearer and easier to follow for devs (easier to understand when we want ErrorSummary to be focused). We can also simplify `formState`.

Are you married to the helper function `focusErrorSummary()`? Without it I think we don't need to assign `errorSummaryRef.current` two places 🧐

Reply:
I found a problem with this method: If ErrorSummary disappears because you fixed all errors, then make a change that makes it show up again, it will be focused.

Reply:
> Are you married to the helper function focusErrorSummary()? Without it I think we don't need to assign errorSummaryRef.current two places 🧐

No can remove that 👍 Since we only focus two places, "rule of three" don't matter here:

Reply:
But at this point, one might just use useEffect without having to navigate resetting refs and more 🙈